### PR TITLE
Fixes for reference replay in discovered issue

### PR DESCRIPTION
**Pull request by csarofeen**
See the added test for the problem case. There were a couple issues in here:
1) Loop and Index mapping could pick concrete domains like `{I0}` instead of `{I0*1}`. Picking the extra broadcast in the concrete domain is actually important for indexing. The added test case fails when selecting `{I0}`
2) When validating we have a valid replay we need to check the index map *or* the loop nest have matching mappings from the loop ID's to the replayed leaf ids. Index map and Loop nest maps can be very different, particularly loop mappings do not necessarily have all root mappings, but index map does. There was an issue in this validation with only looking at loop nests but picking concrete domains with breaking ties on broadcast. I think this just exposed the issue, I don't think it generated it as randomly the domains could be selected as all preferring ones with maximum concrete ID count and maximum broadcast ID count.

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -5708,6 +5708,44 @@ TEST(NVFuserTest, FusionAdvancedLowering4_CUDA) {
...
+
+  TensorView* tv1 = makeConcreteTensor({5, 3});
+  fusion.addInput(tv1);
+
+  auto tv2 = broadcast(tv1, {false, true, false});
+
+  auto tv3 = add(tv0, tv2);
+
+  fusion.addOutput(tv3);
+  fusion.printMath();
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Needs to be removed.

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 2]
```diff
@@ -12,49 +12,61 @@ namespace fuser {
...
+
+    std::unordered_map<IterDomain*, std::pair<int, int>> count_map;
+    for (auto entry : counter.domain_set_) {
+      auto id = entry.first;
+      auto input_id_set = entry.second;
+      int concrete_counts = 0;
+      int broadcast_counts = 0;
+      for (auto input_id : input_id_set) {
+        if (input_id->isBroadcast() ||
+            (gpu_lower &&
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp:44]: Can `gpu_lower` ever be null? It seems it should always be non-null. If so, shouldn't we assert it's non-null instead of allowing it to be null?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp:44]: Sorry, I forgot `ComputeAtMap` is also used before lowering, so `gpu_lower` can be null.

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 3]
```diff
@@ -98,12 +108,11 @@ class ConcreteInputCounter : public IterVisitor {
...
     for (auto output_id : ir_utils::filterByType<IterDomain>(expr->outputs())) {
-      concrete_domain_set_.emplace(std::make_pair(output_id, resulting_set));
+      domain_set_.emplace(std::make_pair(output_id, resulting_set));
     }
   }
 
-  std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>>
-      concrete_domain_set_;
+  std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>> domain_set_;
   GpuLower* gpu_lower_ = nullptr;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: Looks like it's no longer used.

File: torch/csrc/jit/codegen/cuda/index_reference_replay.cpp [hunk 4]
```diff
@@ -217,8 +217,16 @@ TensorDomain* IndexReferenceReplay::computeReplay() {
...
+          // map and loop map do not have the same concrete id mapping. We also
+          // allow matching explicitly through the index map. Index map is not
+          // gauranteed to be contained in loop map, therefore if we generate
+          // mappings to conrete id's through the index map, the mapping from
+          // those ID's to the ID's we replay are not gauranteed to be in loop
+          // map. The reverse is also true, so for validation make sure one of
+          // the mappings exist. For reference check the difference between:
+          // AdvancedLowering5 test and AdvancedIndexing1.
+          if (gpu_lower->caLoopMap().areMapped(id, loop_id) ||
+              gpu_lower->caIndexMap().areMapped(id, loop_id)) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_reference_replay.cpp:229]: It took me a while to understand why this change is needed. 

IIUC, only time we also need to look at the index map is `id` and `loop_id` are root axes. The loop map overwrites a mapping for broadcast input axes when forwarding happens. This isn't a problem when `loop_id` is not root , but when it is, the loop map may not have a mapping from `loop_id` to the concrete ID chosen from the index map.

**Review by naoyam — APPROVED** [review]
